docs/RECIPES.md

```diff
@@ -137,3 +137,34 @@ EOF
 ```shell
 oc apply -f preflight.yaml
 ```
+
+## Container Policy
+These examples are shown using the Container policy against a container image
+(e.g. `preflight check container <image>`). Container policy only runs as a binary on your workstation. Check the latest
+[release](https://github.com/redhat-openshift-ecosystem/openshift-preflight/releases) for the binary that matches your operating system.
+
+You will also need:
+- Your container image published to a container registry
+- A Certification Project ID of the project that was set up in Red Hat Partner Connect
+- A Partner Connect API Key
+- Password to run `sudo` for some container policy checks (`preflight` will prompt you to enter for a given check)
+
+### Testing a Container
+Running container policy checks against a container iteratively until all tests pass.
+
+```bash
+preflight check container registry.example.org/your-namespace/your-image:sometag \
+--pyxis-api-token=abcdefghijklmnopqrstuvwxyz123456 \
+--certification-project-id=1234567890a987654321bcde 
+```
+
+### Submitting a Container's Test Results to Red Hat
+Running container policy checks against a container that has passed all tests and results need to be submitted to Red Hat.
+
+```bash
+preflight check container registry.example.org/your-namespace/your-image:sometag \
+--submit \
+--pyxis-api-token=abcdefghijklmnopqrstuvwxyz123456 \
+--certification-project-id=1234567890a987654321bcde \
+--docker-config=/path/to/your/dockerconfig 
```

Comment:
Hi @bcrochet thank you for the PR, in DCI I'm moving the integration to the latest preflight pre-release, and it's perfect timing.

Could you please add some more examples on how to run preflight using podman, or it's not a recommended way anymore? 
Here are my tests this morning:
```
$ tree
.
├── artifacts
├── kubeconfig
│   └── kubeconfig
└── repoconfig
    └── config.json

# works fine
$ podman run \
    -it \
    --rm \
    --pull=always \
    --name preflight \
    --privileged \
    -e PFLT_JUNIT=true \
    -e PFLT_ARTIFACTS=/artifacts \
    -e PFLT_LOGFILE=/artifacts/preflight.log \
    -e PFLT_LOGLEVEL=trace \
    -v $PWD/artifacts:/artifacts \
    quay.io/opdev/preflight:1.1.0-alpha2 \
    check container \
    quay.io/opdev/simple-demo-operator-bundle:v0.0.3 \
    --certification-project-id my_nice_id \
    --pyxis-api-token my_nice_token

# added docker-config, works fine
$ podman run \
    -it \
    --rm \
    --pull=always \
    --name preflight \
    --privileged \
    -e DOCKER_CONFIG=/repoconfig \
    -e PFLT_JUNIT=true \
    -e PFLT_ARTIFACTS=/artifacts \
    -e PFLT_LOGFILE=/artifacts/preflight.log \
    -e PFLT_LOGLEVEL=trace \
    -v $PWD/artifacts:/artifacts \
    -v $PWD/repoconfig/config.json:/repoconfig/config.json \
    quay.io/opdev/preflight:1.1.0-alpha2 \
    check container \
    quay.io/opdev/simple-demo-operator-bundle:v0.0.3 \
    --certification-project-id my_nice_id \
    --pyxis-api-token my_nice_token \
    --docker-config /repoconfig/config.json
```

Now I simply add `--submit` at the end of the last command, and have an error that I don't understand. Could you please add more detailed examples in the docs? 
```
$ podman run \
    -it \
    --rm \
    --pull=always \
    --name preflight \
    --privileged \
    -e DOCKER_CONFIG=/repoconfig \
    -e PFLT_JUNIT=true \
    -e PFLT_ARTIFACTS=/artifacts \
    -e PFLT_LOGFILE=/artifacts/preflight.log \
    -e PFLT_LOGLEVEL=trace \
    -v $PWD/artifacts:/artifacts \
    -v $PWD/repoconfig/config.json:/repoconfig/config.json \
    quay.io/opdev/preflight:1.1.0-alpha2 \
    check container \
    quay.io/opdev/simple-demo-operator-bundle:v0.0.3 \
    --certification-project-id my_nice_id \
    --pyxis-api-token my_nice_token \
    --docker-config /repoconfig/config.json \
    --submit
Trying to pull quay.io/opdev/preflight:1.1.0-alpha2...
Getting image source signatures
Copying blob 81aea7b17b0e skipped: already exists  
Copying blob 0c9ea4103699 skipped: already exists  
Copying blob 0672ccd2448e skipped: already exists  
Copying config f47f7508a0 done  
Writing manifest to image destination
Storing signatures
time="2022-03-01T11:09:37Z" level=info msg="config file not found, proceeding without it"
time="2022-03-01T11:09:37Z" level=info msg="certification library version 1.1.0-alpha2 <commit: 45d66d1d824423f1655ab6fb8fb5e892f83ec9c2>"
time="2022-03-01T11:09:37Z" level=debug msg="URL is: https://catalog.redhat.com/api/containers/v1/projects/certification/id/61a64ad87a3f492d8b755778"
time="2022-03-01T11:09:38Z" level=debug msg="Certification project name is: Test Project: certify simple-demo-operator"
time="2022-03-01T11:09:38Z" level=info msg="target image: quay.io/opdev/simple-demo-operator-bundle:v0.0.3"
time="2022-03-01T11:09:38Z" level=info msg="pulling image from target registry"
time="2022-03-01T11:09:39Z" level=debug msg="temporary directory is /tmp/preflight-2281296111"
time="2022-03-01T11:09:39Z" level=debug msg="exporting and flattening image"
time="2022-03-01T11:09:39Z" level=debug msg="extracting container filesystem to /tmp/preflight-2281296111/fs"
time="2022-03-01T11:09:39Z" level=debug msg="writing container filesystem to output dir: /tmp/preflight-2281296111/fs"
time="2022-03-01T11:09:39Z" level=trace msg="image config written to disk: /artifacts/cert-image.json"
Error: open /tmp/preflight-2281296111/fs/var/lib/rpm/Packages: no such file or directory
time="2022-03-01T11:09:39Z" level=fatal msg="open /tmp/preflight-2281296111/fs/var/lib/rpm/Packages: no such file or directory"
```

Reply:
@tkrishtop One more thing I just realized, you are passing in a `bundle` image, this will not work for container checks. The container needs to be a real container based on RHEL that is passed in for `check container`.

Reply:
@acornett21 thank you, I have a lot of questions now, will ask them in the slack channel.

Reply:
@acornett21 I'm trying to run preflight binary as described in this docs:

```
$ DOCKER_CONFIG=$PWD/repoconfig \
PFLT_JUNIT=true \
PFLT_ARTIFACTS=$PWD/artifacts \
PFLT_LOGFILE=$PWD/artifacts/preflight.log \
PFLT_LOGLEVEL=trace \
preflight \
check container \
quay.io/opdev/simple-demo-operator:0.0.3 \
--certification-project-id=my_nice_id \
--pyxis-api-token=my_nice_token \
--docker-config=$PWD/repoconfig/config.json \
--submit 
```

Got three issues:

- Preflight asks for sudo password in the middle of the execution. It seems like the invocation of preflight binary should be done with sudo
- I provide docker-config but still have a warning `config file not found, proceeding without it`. Is there another config to provide?
- Submit-to-Pyxis part fails with unclear error (I'll open an issue).

Here are detailed logs: 
```
$ DOCKER_CONFIG=$PWD/repoconfig \
PFLT_JUNIT=true \
PFLT_ARTIFACTS=$PWD/artifacts \
PFLT_LOGFILE=$PWD//artifacts/preflight.log \
PFLT_LOGLEVEL=trace \
preflight \
check container \
quay.io/opdev/simple-demo-operator:0.0.3 \
--certification-project-id=61a64ad87a3f492d8b755778 \
--pyxis-api-token=my_nice_token \
--docker-config=$PWD/repoconfig/config.json \
--submit 
time="2022-03-02T13:09:04+01:00" level=info msg="config file not found, proceeding without it"
time="2022-03-02T13:09:04+01:00" level=info msg="certification library version 0.0.0 <commit: 45d66d1d824423f1655ab6fb8fb5e892f83ec9c2>"
time="2022-03-02T13:09:04+01:00" level=debug msg="URL is: https://catalog.redhat.com/api/containers/v1/projects/certification/id/61a64ad87a3f492d8b755778"
time="2022-03-02T13:09:05+01:00" level=debug msg="Certification project name is: Test Project: certify simple-demo-operator"
time="2022-03-02T13:09:05+01:00" level=info msg="target image: quay.io/opdev/simple-demo-operator:0.0.3"
time="2022-03-02T13:09:05+01:00" level=info msg="pulling image from target registry"
time="2022-03-02T13:09:06+01:00" level=debug msg="temporary directory is /tmp/preflight-1286440083"
time="2022-03-02T13:09:06+01:00" level=debug msg="exporting and flattening image"
time="2022-03-02T13:09:06+01:00" level=debug msg="extracting container filesystem to /tmp/preflight-1286440083/fs"
time="2022-03-02T13:09:06+01:00" level=debug msg="writing container filesystem to output dir: /tmp/preflight-1286440083/fs"
time="2022-03-02T13:09:12+01:00" level=trace msg="image config written to disk: /home/tkrishto/Documents/test_field/artifacts/cert-image.json"
time="2022-03-02T13:09:12+01:00" level=trace msg="rpm manifest written to disk: /home/tkrishto/Documents/test_field/artifacts/rpm-manifest.json"
time="2022-03-02T13:09:12+01:00" level=info msg="Container checks do not require a cluster. skipping cluster version check."
time="2022-03-02T13:09:12+01:00" level=info msg="executing checks"
time="2022-03-02T13:09:12+01:00" level=info msg="running check: HasUniqueTag"
time="2022-03-02T13:09:12+01:00" level=info msg="check completed: HasUniqueTag" result=PASSED
time="2022-03-02T13:09:12+01:00" level=info msg="running check: LayerCountAcceptable"
time="2022-03-02T13:09:12+01:00" level=debug msg="detected 3 layers in image"
time="2022-03-02T13:09:12+01:00" level=info msg="check completed: LayerCountAcceptable" result=PASSED
time="2022-03-02T13:09:12+01:00" level=info msg="running check: RunAsNonRoot"
time="2022-03-02T13:09:12+01:00" level=debug msg="User specified that was not root"
time="2022-03-02T13:09:12+01:00" level=info msg="check completed: RunAsNonRoot" result=PASSED
time="2022-03-02T13:09:12+01:00" level=info msg="running check: BasedOnUbi"
time="2022-03-02T13:09:12+01:00" level=trace msg="Has RHEL Name"
time="2022-03-02T13:09:12+01:00" level=trace msg="Has RHEL ID"
time="2022-03-02T13:09:12+01:00" level=info msg="check completed: BasedOnUbi" result=PASSED
time="2022-03-02T13:09:12+01:00" level=info msg="running check: HasLicense"
time="2022-03-02T13:09:12+01:00" level=error msg="Error when checking for /licenses : stat /tmp/preflight-1286440083/fs/licenses: no such file or directory"
time="2022-03-02T13:09:12+01:00" level=info msg="check completed: HasLicense" result=FAILED
time="2022-03-02T13:09:12+01:00" level=info msg="running check: HasNoProhibitedPackagesMounted"
time="2022-03-02T13:09:12+01:00" level=info msg="check completed: HasNoProhibitedPackagesMounted" result=PASSED
time="2022-03-02T13:09:12+01:00" level=info msg="running check: HasRequiredLabel"
time="2022-03-02T13:09:12+01:00" level=info msg="check completed: HasRequiredLabel" result=PASSED
time="2022-03-02T13:09:12+01:00" level=info msg="running check: RunnableContainer"
time="2022-03-02T13:09:12+01:00" level=debug msg="Creating container quay.io/opdev/simple-demo-operator:0.0.3 with the run options: {Entrypoint:[] Cmd:[sleep 10s] ContainerName:}"
time="2022-03-02T13:09:12+01:00" level=debug msg="Command being run: /usr/bin/podman create quay.io/opdev/simple-demo-operator:0.0.3 sleep 10s"
time="2022-03-02T13:09:14+01:00" level=debug msg="Successfully created container quay.io/opdev/simple-demo-operator:0.0.3 with options {Entrypoint:[] Cmd:[sleep 10s] ContainerName:} ..."
time="2022-03-02T13:09:14+01:00" level=debug msg="Container Id is 1bfbd90ffb6fd6aa02db6a97954a7b31c05ca4119c54b648de5372cafcaba087"
time="2022-03-02T13:09:14+01:00" level=debug msg="Starting container 1bfbd90ffb6fd6aa02db6a97954a7b31c05ca4119c54b648de5372cafcaba087"
time="2022-03-02T13:09:14+01:00" level=debug msg="Command being run: /usr/bin/podman start 1bfbd90ffb6fd6aa02db6a97954a7b31c05ca4119c54b648de5372cafcaba087"
time="2022-03-02T13:09:14+01:00" level=debug msg="Successfully started container 1bfbd90ffb6fd6aa02db6a97954a7b31c05ca4119c54b648de5372cafcaba087..."
time="2022-03-02T13:09:14+01:00" level=debug msg="Checking for the status of the container 1bfbd90ffb6fd6aa02db6a97954a7b31c05ca4119c54b648de5372cafcaba087..."
time="2022-03-02T13:09:14+01:00" level=debug msg="Command being run: /usr/bin/podman wait --interval 2s --condition running 1bfbd90ffb6fd6aa02db6a97954a7b31c05ca4119c54b648de5372cafcaba087"
time="2022-03-02T13:09:24+01:00" level=error msg="unable to wait for container 1bfbd90ffb6fd6aa02db6a97954a7b31c05ca4119c54b648de5372cafcaba087: signal: killed"
time="2022-03-02T13:09:24+01:00" level=error msg="Stderr: "
time="2022-03-02T13:09:24+01:00" level=debug msg="Removing container 1bfbd90ffb6fd6aa02db6a97954a7b31c05ca4119c54b648de5372cafcaba087"
time="2022-03-02T13:09:24+01:00" level=debug msg="Command being run: /usr/bin/podman rm --force 1bfbd90ffb6fd6aa02db6a97954a7b31c05ca4119c54b648de5372cafcaba087"
time="2022-03-02T13:09:24+01:00" level=info msg="check completed: RunnableContainer" result=FAILED
time="2022-03-02T13:09:24+01:00" level=info msg="running check: RunSystemContainer"
time="2022-03-02T13:09:24+01:00" level=debug msg="Creating container quay.io/opdev/simple-demo-operator:0.0.3 with the run options: {Entrypoint:[] Cmd:[] ContainerName:podman-test}"
time="2022-03-02T13:09:24+01:00" level=debug msg="Command being run: /usr/bin/podman create --name podman-test quay.io/opdev/simple-demo-operator:0.0.3"
time="2022-03-02T13:09:25+01:00" level=debug msg="Successfully created container quay.io/opdev/simple-demo-operator:0.0.3 with options {Entrypoint:[] Cmd:[] ContainerName:podman-test} ..."
time="2022-03-02T13:09:25+01:00" level=debug msg="Container Id is 9e162831eb8ffbd5e95319a17e584fb811afc9cff94ca80f071ab1f1a642b721"
time="2022-03-02T13:09:25+01:00" level=debug msg="Generating Systemd file for container podman-test"
time="2022-03-02T13:09:25+01:00" level=debug msg="Command being run: /usr/bin/podman generate systemd --new --files --name podman-test"
time="2022-03-02T13:09:25+01:00" level=debug msg="Copying /home/tkrishto/Documents/test_field/container-podman-test.service into /etc/systemd/system"
time="2022-03-02T13:09:25+01:00" level=debug msg="Command being run: /usr/bin/sh -c sudo cp /home/tkrishto/Documents/test_field/container-podman-test.service /etc/systemd/system"
[sudo] password for tkrishto: 
time="2022-03-02T13:09:42+01:00" level=debug msg="Successfully copied /home/tkrishto/Documents/test_field/container-podman-test.service to /etc/systemd/system"
time="2022-03-02T13:09:42+01:00" level=debug msg="Reloading daemon set and start the service container-podman-test.service"
time="2022-03-02T13:09:42+01:00" level=debug msg="Command being run: /usr/bin/sh -c sudo systemctl daemon-reload"
time="2022-03-02T13:09:42+01:00" level=debug msg="Command being run: /usr/bin/sh -c sudo systemctl start container-podman-test.service"
time="2022-03-02T13:09:45+01:00" level=debug msg="Checking the service status of container-podman-test.service"
time="2022-03-02T13:09:45+01:00" level=debug msg="Command being run: /usr/bin/sh -c sudo systemctl is-active container-podman-test.service"
time="2022-03-02T13:09:45+01:00" level=debug msg="The container-podman-test.service status is active"
time="2022-03-02T13:09:45+01:00" level=debug msg="Stopping the container service container-podman-test.service"
time="2022-03-02T13:09:45+01:00" level=debug msg="Command being run: /usr/bin/sh -c sudo systemctl stop container-podman-test.service"
time="2022-03-02T13:09:46+01:00" level=debug msg="Command being run: /usr/bin/sh -c sudo systemctl daemon-reload"
time="2022-03-02T13:09:46+01:00" level=debug msg="Successfully stopped the container service container-podman-test.service"
time="2022-03-02T13:09:46+01:00" level=debug msg="Removing container podman-test"
time="2022-03-02T13:09:46+01:00" level=debug msg="Command being run: /usr/bin/podman rm --force podman-test"
time="2022-03-02T13:09:46+01:00" level=info msg="check completed: RunSystemContainer" result=PASSED
{
    "image": "quay.io/opdev/simple-demo-operator:0.0.3",
    "passed": false,
    "test_library": {
        "name": "github.com/redhat-openshift-ecosystem/openshift-preflight",
        "version": "0.0.0",
        "commit": "45d66d1d824423f1655ab6fb8fb5e892f83ec9c2"
    },
    "results": {
        "passed": [
            {
                "name": "HasUniqueTag",
                "elapsed_time": 583,
                "description": "Checking if container has a tag other than 'latest', so that the image can be uniquely identified."
            },
            {
                "name": "LayerCountAcceptable",
                "description": "Checking if container has less than 40 layers.  Too many layers within the container images can degrade container performance."
            },
            {
                "name": "RunAsNonRoot",
                "description": "Checking if container runs as the root user because a container that does not specify a non-root user will fail the automatic certification, and will be subject to a manual review before the container can be approved for publication"
            },
            {
                "name": "BasedOnUbi",
                "description": "Checking if the container's base image is based upon the Red Hat Universal Base Image (UBI)"
            },
            {
                "name": "HasNoProhibitedPackagesMounted",
                "elapsed_time": 20,
                "description": "Checks to ensure that the image in use does not include prohibited packages, such as Red Hat Enterprise Linux (RHEL) kernel packages."
            },
            {
                "name": "HasRequiredLabel",
                "description": "Checking if the required labels (name, vendor, version, release, summary, description) are present in the container metadata."
            },
            {
                "name": "RunSystemContainer",
                "elapsed_time": 21935,
                "description": "Checking if a container can run as a systemd service"
            }
        ],
        "failed": [
            {
                "name": "HasLicense",
                "description": "Checking if terms and conditions applicable to the software including open source licensing information are present. The license must be at /licenses",
                "help": "Check HasLicense encountered an error. Please review the preflight.log file for more information.",
                "suggestion": "Create a directory named /licenses and include all relevant licensing and/or terms and conditions as text file(s) in that directory.",
                "knowledgebase_url": "https://connect.redhat.com/zones/containers/container-certification-policy-guide",
                "check_url": "https://connect.redhat.com/zones/containers/container-certification-policy-guide"
            },
            {
                "name": "RunnableContainer",
                "elapsed_time": 12245,
                "description": "Checking if container runs within a pre-configured timeframe",
                "help": "Check RunnableContainer encountered an error. Please review the preflight.log file for more information.",
                "suggestion": "Ensure that the container can be launched",
                "knowledgebase_url": "https://connect.redhat.com/zones/containers/container-certification-policy-guide",
                "check_url": "https://connect.redhat.com/zones/containers/container-certification-policy-guide"
            }
        ],
        "errors": []
    }
}
time="2022-03-02T13:09:46+01:00" level=trace msg="JUnitXML written to /home/tkrishto/Documents/test_field/artifacts/results-junit.xml"
time="2022-03-02T13:09:46+01:00" level=info msg="preparing results that will be submitted to Red Hat"
time="2022-03-02T13:09:47+01:00" level=error msg="received non 200 status code in create manifest call: {\"detail\": \"Additional properties are not allowed ('raw_config' was unexpected)\", \"status\": 400, \"title\": \"Bad Request\", \"type\": \"about:blank\", \"trace_id\": \"0x1eba7d8da5c4f2f95b56d3b1e236657f\"}"
time="2022-03-02T13:09:47+01:00" level=error msg="error calling remote APIcould not create image"
Error: error calling remote API
time="2022-03-02T13:09:47+01:00" level=fatal msg="error calling remote API"


```

Reply:
> * Preflight asks for sudo password in the middle of the execution. It seems like the invocation of preflight binary should be done with sudo
> * I provide docker-config but still have a warning `config file not found, proceeding without it`. Is there another config to provide?
> * Submit-to-Pyxis part fails with unclear error (I'll open an issue).

@tkrishtop Below are my thoughts:
1. For some of the checks in the build you are testing need `sudo` we did not want to have a user run `sudo` for all checks, and went with a `minimum` access approach, hence the prompt before the check is executed. This check (and others that require sudo, as well as other checks) are still in flux, and have been removed in `main`.
2. That message is if you are using `config.yaml` to run preflight itself, but you are using `env's` directly. This is an either or situation, either use a `config.yaml` with all of the `PFLT_` env's, or pass them in like you are.
3. This error from `pyxis` is because `raw_config` is not exposed in the `external` pyxis API, it is currently only in the `internal` API. They have made this change to the `external` api as well, but that change is only in QA at this point. 

Let me know if something does not make sense. Thanks for all your help testing this.

Reply:
The problem with prompting in the middle of the run is that it makes running Preflight in a CI/CD environment harder, since the harness has to be able to deal with being asked for a password. Better in that case to require running it with `sudo` if some tests require elevated privileges. This is similar to the `oscap-podman` tool that requires running in rooted mode. 

As to the `config.yaml` question, I would think that the right thing to do here is to have a multi-layered approach as other utilities do, where each of the possible sources of option values has a different weight from lowest to highest:

- Lowest: Internal defaults 
- Next: `config.yaml` options
- Highest: command line options

Each option would be independent of every other.  Putting an option, say `--foo-option=value-a` on the command line would not prevent the option `--bar-option=true` in the config.yaml file from being accepted and used, or the default for option `--baz-option=true` from being applied if it is not mentioned in either the `config.yaml` file or the command line. However, if `--foo-option="value-a"` is included on the command line, and `--foo-option="value-b"` is in the config.yaml file, and `--foo-option="none"` is the default, then `--foo-option="value-a"` would win.

This allows completely predictable behavior, assuming that the order is documented. It also allows people designing a CI workflow to put together 'general' `config.yaml` files that contain common non-standard option and environment definitions required for all of the partner's images and projects, while allowing specific modifications to those common definitions to be specified on the command line.

Reply:
@jfrancin We do not want to run all of `preflight` checks with `sudo` that would be problematic, and give some checks more access than they need. We do not want to give rooted access to preflight. Like I mentioned before, the checks that require `sudo` might go away fully (they are currently disabled in main).

As far as `config.yaml` goes we handle it exactly how you describe, if a user gives preflight a config, it will use it, if you overwrite a value with an `env` we will use that.